devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
+                # If I have been passed a parameter, I must have seen it before
+                if len(orig_param_l) == 0:
+                    raise InvalidArgument("Parameter %s does not exist in expressions " +
+                                          "passed to this Operator" % k)
+                # We've made sure the list isn't empty. Names should be unique so it
+                # should have exactly one entry
+                assert(len(orig_param_l) == 1)
+                orig_param = orig_param_l[0]
+                # Pull out the children and add them to kwargs
+                for orig_child, new_child in zip(orig_param.children, v.children):
+                    new_params[orig_child.name] = new_child
+        kwargs.update(new_params)
+        return kwargs
+
+    def _dle_arguments(self, dim_sizes):
+        # Add user-provided block sizes, if any
+        dle_arguments = OrderedDict()
+        autotune = True
+        for i in self.dle_arguments:
+            dim_size = dim_sizes.get(i.original_dim.name, None)
+            if dim_size is None:
+                error('Unable to derive size of dimension %s from defaults. '
+                      'Please provide an explicit value.' % i.original_dim.name)
+                raise InvalidArgument('Unknown dimension size')
+            if i.value:
+                try:
+                    dle_arguments[i.argument.name] = i.value(dim_size)
+                except TypeError:
+                    dle_arguments[i.argument.name] = i.value
+                    autotune = False
+            else:
+                dle_arguments[i.argument.name] = dim_size
+        return dle_arguments, autotune
+
+    def _derive_values(self, kwargs):
+        # Use kwargs
+        values = OrderedDict()
+        dimension_values = OrderedDict()
+        
+        for i in self.arguments:
+            values[i] = get_value(i, kwargs.pop(i.name, None), values)
+
+        for i in self.dims:
+            dimension_values[i] = kwargs.pop(i.name, None)
+
+        # Make sure we've used all arguments passed
+        if len(kwargs) > 0:
+            raise InvalidArgument("Unknown arguments passed: " + ", ".join(kwargs.keys()))
+
+        # Derive values for other arguments
+        for i in self.arguments:
+            if values[i] is None:
+                known_values = OrderedDict(chain(values.items(), dimension_values.items()))
+                provided_values = [get_value(i, x, known_values) for x in i.gets_value_from]
+                assert(len(provided_values) > 0)
+        
+                if len(provided_values) == 1:
+                    values[i] = provided_values[0]
+                else:
+                    values[i] = reduce(provided_values)
+        
+        # Second pass to evaluate any Unevaluated dependencies from the first pass
+        for k, v in values.items():
+            if isinstance(v, UnevaluatedDependency):
+                values[k] = v.evaluate(values)
```

Comment:
not so pythonic, but most importantly changing a dict in place through ``items`` which is not doable, I seem to remember, in P3 (correct me if I'm wrong?)

potential fix:
``values = OrderedDict([(k, v.evaluate(values) if isinstance(v, UnevaluatedDependency) else v) for k, v in values.items()])``

Alternative: ``for k, v list(values.items()): ...``

Reply:
I believe this is working. I don't see what's wrong. The line that changes the dict does not access it using items, it does it using `values[k]` so it should just work, and I believe it does.

Reply:
weird... 
anyway, if the dictionary isn't huge, a more pyhtonic way might help:

``values = {k: v.evaluate(values) if isinstance(v, UnevaluatedDependency) else v for k, v in values.items()}``